src/rust/protocol/redis/src/request/get.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use super::*;
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Get {
+    key: Box<[u8]>,
+}
+
+impl Get {
+    pub fn key(&self) -> &[u8] {
+        &self.key
+    }
+}
+
+impl RequestParser {
+    // this is to be called after parsing the command, so we do not match the verb
+    pub(crate) fn parse_get_no_stats<'a>(&self, input: &'a [u8]) -> IResult<&'a [u8], Get> {
+        let (input, _) = space1(input)?;
+        let (input, key) = string(input, self.max_key_len)?;
+        let (input, _) = space0(input)?;
+        let (input, _) = crlf(input)?;
+        Ok((
+            input,
+            Get {
+                key: key.to_owned().into_boxed_slice(),
+            },
+        ))
+    }
+
+    // this is to be called after parsing the command, so we do not match the verb
+    pub fn parse_get<'a>(&self, input: &'a [u8]) -> IResult<&'a [u8], Get> {
+        match self.parse_get_no_stats(input) {
+            Ok((input, request)) => {
+                GET.increment();
+                GET_KEY.increment();
+                Ok((input, request))
+            }
+            Err(e) => {
+                if !e.is_incomplete() {
+                    GET.increment();
+                    GET_EX.increment();
+                }
+                Err(e)
+            }
+        }
+    }
+}
+
+impl Compose for Get {
+    fn compose(&self, session: &mut session::Session) {
```

Comment:
That's a good point. I skipped over the client side stats in the memcache protocol so far as well. Tough to choose between the client/server and compose/parse pairs - I can see merit to both. This likely requires some extension to the metrics macros to allow us to specify a namespace.

Reply:
Personally I prefer `parse`/`compose` because the meaning is in tune with the scope of this module, which is just for protocols but not the processing that follows. Imagine we build a cache proxy with a backend of 2 replicas, where we call client of this proxy as upstream and backend as downstream, here's what request flow looks like if the first backend didn't have the key we want but its mirror does:

1. parse upstream request,
2. compose downstream request,
3. send to one backend server (say `A`),
4. parse server response from `A` which turns out to be a miss,
5. send the same downstream request to the other backend server (say `B`),
6. parse server response from `B` which turns out to be a hit,
7. compose upstream response,
8. send response back to upstream.

In this example, we parsed 1 upstream request, 2 downstream responses, and composed 1 downstream request and 1 upstream response, and created 2 request RPCs as well as 1 response RPC.

When it comes to metrics, I think we will need a few namespaces to describe this flow: `parse` (`parse/get: 1`, `parse/get/key_miss: 1`, `parse/get/key_hit: 1`), `compose` (`compose/get: 1`, `compose/get/key_hit: 1`) for the protocol related activity, as well as `clnt` (`clnt/send: 2`, `clnt/recv: 2`) and `server`  (`server/recv: 1`, `server/send: 1`) that captures RPCs.

Reply:
maybe this is why most codec has no metrics at all ...

Reply:
Haha. I still think we can do something here. It's just going to take some thought to make sure the ergonomics are good and the metrics give us the right flexibility. Things seem to get a bit worse if we consider how Prometheus/opentelemetry metrics tend to work...

Reply:
What aspects of opentelemetry make this more harder/worse?

Reply:
Take a look at the prometheus metric naming guide, particularly around the recommendations for labels: https://prometheus.io/docs/practices/naming

Reply:
Let's just say I'm not impressed:

> `process_cpu_seconds_total` 
I'm sorry- what? Do you want the unit at the end or not? 

I'm gonna discard this naming guide, it'll never prevail against common sense.

Reply:
Haha. Yeah, there are some curious suggestions there. I have no objection to disregarding that and rolling with whatever seems sensible.